Client/Frontend/Login Management/BreachAlertsManager.swift

```diff
@@ -30,50 +30,66 @@ final public class BreachAlertsManager {
+            return
+        }
+
+        // 1b. local file exists, so load from that
+        guard let fileData = FileManager.default.contents(atPath: self.cacheURL.path) else {
+            completion(Maybe(failure: BreachAlertsError(description: "failed to get data from breach.json")))
+            Sentry.shared.send(message: "BreachAlerts: failed to get data from breach.json")
+            try? FileManager.default.removeItem(at: self.cacheURL) // bad file, so delete it
+            self.fetchAndSaveBreaches(completion)
+            return
+        }
+
+        // 2. check the last time breach endpoint was accessed
+        guard let dateLastAccessed = profile.prefs.timestampForKey(BreachAlertsClient.etagDateKey) else {
+            profile.prefs.removeObjectForKey(BreachAlertsClient.etagDateKey) // bad key, so delete it
+            self.fetchAndSaveBreaches(completion)
+            return
+        }
+        let timeUntilNextUpdate = UInt64(60 * 60 * 24 * 3 * 1000) // 3 days in milliseconds
+        let shouldUpdateDate = dateLastAccessed + timeUntilNextUpdate
+
+        // 3. if 3 days have not passed since last update...
+        guard Date.now() >= shouldUpdateDate else {
+            // 3a. no need to refetch. decode local data and hand off
+            decodeData(data: fileData, completion)
+            return
+        }
+
+        // 3b. should update - check if the etag is different
+        client.fetchEtag(endpoint: .breachedAccounts, profile: self.profile) { etag in
+            guard let etag =  etag else {
+                self.profile.prefs.removeObjectForKey(BreachAlertsClient.etagKey) // bad key, so delete it
                 return
             }
+            let savedEtag = self.profile.prefs.stringForKey(BreachAlertsClient.etagKey)
 
-            self.breaches = decoded
-            // remove for release
-            self.breaches.insert(BreachRecord(
-             name: "MockBreach",
-             title: "A Mock Blockbuster Record",
-             domain: "blockbuster.com",
-             breachDate: "1970-01-02",
-             description: "A mock BreachRecord for testing purposes."
-            ))
-            self.breaches.insert(BreachRecord(
-             name: "MockBreach",
-             title: "A Mock Lorem Ipsum Record",
-             domain: "lipsum.com",
-             breachDate: "1970-01-02",
-             description: "A mock BreachRecord for testing purposes."
-            ))
-            self.breaches.insert(BreachRecord(
-             name: "MockBreach",
-             title: "A Mock Swift Breach Record",
-             domain: "swift.org",
-             breachDate: "1970-01-02",
-             description: "A mock BreachRecord for testing purposes."
-            ))
-            completion(Maybe(success: self.breaches))
+            // 4. if it is, refetch the data and hand entire Set of BreachRecords off
+            if etag != savedEtag {
+                self.fetchAndSaveBreaches(completion)
+            }
```

Comment:
```
else { 
update the BreachAlertsClient.etagDateKey
decodeData(data: fileData, completion)
}
``` 
Or did that happen elsewhere> I am trying to understand the case where >3 days and the etag is unchanged. I assume update the date, and decode the existing file.

Reply:
You're right, as it is it will fall through and nothing will happen. Added the else case